application/src/main/java/org/togetherjava/tjbot/commands/tags/TagManageCommand.java

```diff
@@ -364,37 +364,39 @@ private void logAction(@NotNull Subcommand subcommand, @NotNull Guild guild,
                     "previousContent is null even though the subcommand should supply a value.");
         }
 
+        BiConsumer<String, ModAuditLogWriter.Attachment[]> writeLog =
+                (title, attachments) -> modAuditLogWriter.write(title,
+                        String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id),
+                        author, triggeredAt, guild, attachments);
+
         switch (subcommand) {
-            case CREATE -> modAuditLogWriter.write("Tag-Manage Create",
-                    String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id), author,
-                    triggeredAt, guild, new ModAuditLogWriter.Attachment(CONTENT_FILE_NAME,
-                            Objects.requireNonNull(newContent)));
-
-            case CREATE_WITH_MESSAGE -> modAuditLogWriter.write("Tag-Manage Create with message",
-                    String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id), author,
-                    triggeredAt, guild, new ModAuditLogWriter.Attachment(CONTENT_FILE_NAME,
-                            Objects.requireNonNull(newContent)));
-
-            case EDIT -> modAuditLogWriter.write("Tag-Manage Edit",
-                    String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id), author,
-                    triggeredAt, guild,
-                    new ModAuditLogWriter.Attachment(NEW_CONTENT_FILE_NAME,
-                            Objects.requireNonNull(newContent)),
-                    new ModAuditLogWriter.Attachment(PREVIOUS_CONTENT_FILE_NAME,
-                            Objects.requireNonNull(previousContent)));
-
-            case EDIT_WITH_MESSAGE -> modAuditLogWriter.write("Tag-Manage Edit with message",
-                    String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id), author,
-                    triggeredAt, guild,
-                    new ModAuditLogWriter.Attachment(NEW_CONTENT_FILE_NAME,
-                            Objects.requireNonNull(newContent)),
-                    new ModAuditLogWriter.Attachment(PREVIOUS_CONTENT_FILE_NAME,
-                            Objects.requireNonNull(previousContent)));
-
-            case DELETE -> modAuditLogWriter.write("Tag-Manage Delete",
-                    String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id), author,
-                    triggeredAt, guild, new ModAuditLogWriter.Attachment(PREVIOUS_CONTENT_FILE_NAME,
-                            Objects.requireNonNull(previousContent)));
+            case CREATE -> writeLog.accept("Tag-Manage Create",
+                    new ModAuditLogWriter.Attachment[] {new ModAuditLogWriter.Attachment(
+                            CONTENT_FILE_NAME, Objects.requireNonNull(newContent))});
```

Comment:
u can get rid of the Objects.requireNonNull in all the cases here. also makes the code slimmer.

Reply:
Sonar doesn't like it. just re `check`ed

![image](https://user-images.githubusercontent.com/63251610/158010925-916cc355-1b64-4bf4-9865-4058a18df4a1.png)

Reply:
the problem is that the linter cant make the connection between `if (Subcommand.SUBCOMMANDS_WITH_NEW_CONTENT.contains(subcommand) && newContent == null) {` which checks for null-safety and this part where u use `newContent` with `CREATE`.

While ur code is actually safe, it is indeed a flaw that sonar is detecting here. Suppose you would add a new enum case `FOO` but forget to add it to either `SUBCOMMANDS_WITH_NEW_CONTENT` or `` SUBCOMMANDS_WITH_PREVIOUS_CONTENT`. Then the code would indeed be incorrect and crash here; but only at runtime.

At this point, I would instead create your attachments outside of the switch-case, based on the two enumsets. You can also move the null-checks inside then. Then the BiConsumer is also obsolete since the switch only decides the title, nothing else:
```java
List<ModAuditLogWriter.Attachment> attachments = new ArrayList<>();
if (SUBCOMMANDS_WITH_NEW_CONTENT.contains(subcommand)) {
  if (newContent == null) {
    throw new IllegalArgumentException(...);
  }

  String fileName = subcommand == CREATE || subcommand == CREATE_WITH_MESSAGE
    ? CONTENT_FILE_NAME
    : NEW_CONTENT_FILE_NAME;

  attachments.add(new ModAuditLogWriter.Attachment(fileName, newContent));
}

if (SUBCOMMANDS_WITH_PREVIOUS_CONTENT.contains(subcommand)) {
  if (previousContent == null) {
    throw new IllegalArgumentException(...);
  }

  attachments.add(new ModAuditLogWriter.Attachment(PREVIOUS_CONTENT_FILE_NAME, previousContent));
}

String title = switch (subcommand) {
  case CREATE -> "Tag-Manage Create"
  case CREATE_WITH_MESSAGE -> "Tag-Manage Create with message"
  ...
};

modAuditLogWriter.write(
  title,
  String.format(LOG_EMBED_DESCRIPTION, subcommand.getActionVerb(), id),
  author, triggeredAt, guild,
  attachments.toArray(Attachment[]::new));
```

Reply:
See 279d3aa68f6aca0068c673beeef94b2dd25083e2